### PR TITLE
adding retry event count metric for source

### PR DESCRIPTION
**Pull request by kthakar1990**
# Changes

- adding a code change that will be used by knative/eventing to report retry event count for knative source.

### CODE REVIEW

File: source/source_stats_reporter.go [hunk 1]
```diff
@@ -51,6 +57,7 @@ var (
 	responseTimeout        = tag.MustNewKey(metricskey.LabelResponseTimeout)
 )
 
+//ReportArgs interface
```
Comment (skonto) [inline source/source_stats_reporter.go:60]: This is a struct, not sure if we want to add this here.

File: source/source_stats_reporter_test.go [hunk 2]
```diff
@@ -51,14 +51,31 @@ func TestStatsReporter(t *testing.T) {
...
 	expectSuccess(t, func() error {
 		return r.ReportEventCount(args, http.StatusAccepted)
 	})
 	expectSuccess(t, func() error {
 		return r.ReportEventCount(args, http.StatusAccepted)
 	})
+	expectSuccess(t, func() error {
+		return r.ReportRetryEventCount(args, http.StatusServiceUnavailable)
+	})
+	expectSuccess(t, func() error {
```
Comment (skonto) [inline source/source_stats_reporter_test.go:74]: You repeat the same function twice?

File: source/source_stats_reporter.go (hunk 1 above)
Comment (kthakar1990) [reply source/source_stats_reporter.go:60]: I can remove it.. it was just giving an suggestion to be exported or commented.. so I added it there

**Comment by kthakar1990** [comment]
United test failures would be resolved by this [PR](https://github.com/knative/eventing/pull/4861)

File: .github/workflows/knative-releasability.yaml [hunk 3]
```diff
@@ -36,8 +36,8 @@ jobs:
     env:
       #########################################
       #   Update this section each release.   #
-      RELEASE: 'v0.20'
-      SLACK_CHANNEL: 'release-20'
+      RELEASE: 'v0.21'
+      SLACK_CHANNEL: 'release-21'
```
Comment (skonto) [inline .github/workflows/knative-releasability.yaml:40]: I think this file is not related to this PR.

File: source/source_stats_reporter.go [hunk 4]
```diff
@@ -51,6 +57,7 @@ var (
 	responseTimeout        = tag.MustNewKey(metricskey.LabelResponseTimeout)
 )
 
+//ReportArgs defines the arguments for reporting metrics
```
Comment (julz) [inline source/source_stats_reporter.go:60]:
```suggestion
// ReportArgs defines the arguments for reporting metrics.
```

File: source/source_stats_reporter.go [hunk 5]
```diff
@@ -35,6 +35,12 @@ var (
 		stats.UnitDimensionless,
 	)
 
+	// RetryEventCountM is a counter which records the number of events sent by the source in retries.
+	RetryEventCountM = stats.Int64(
```
Comment (julz) [inline source/source_stats_reporter.go:39]: just to check: does this need to be exported? (I notice `eventCountM` isn't and I'm wondering if this can be private too)

Comment (kthakar1990) [reply source/source_stats_reporter.go]: you're right it doesn't have to be!

**Review by julz — COMMENTED** [review]
/lgtm

**Comment by evankanderson** [comment]
Just wondering if you want to use a distribution of retry counts to distinguish between "everything retries once" and "20% retry 5 times".

**Comment by evankanderson** [comment]
Also, `pkg` seems like an unfortunate place for us to need to do development for sources. I don't have a pat solution, though.

**Comment by vagababov** [comment]
> Also, `pkg` seems like an unfortunate place for us to need to do development for sources. I don't have a pat solution, though.

Yeah, I think I mentioned that in one other PR.
Would just eventing be a good place? DOn't sources depend on it anyway?

**Comment by kthakar1990** [comment]
> Just wondering if you want to use a distribution of retry counts to distinguish between "everything retries once" and "20% retry 5 times".

What we actually want is to see various count by http status codes.. so retry count is enough for us.. because we can segment by http status codes..

**Comment by kthakar1990** [comment]
> > Also, `pkg` seems like an unfortunate place for us to need to do development for sources. I don't have a pat solution, though.
> 
> Yeah, I think I mentioned that in one other PR.
> Would just eventing be a good place? DOn't sources depend on it anyway?

I am fine with that idea.. However, there is a plan to make standardize metrics API and we can include the refactoring with that work..

**Comment by vaikas** [comment]
I don't have a super strong feeling about this landing in pkg, just wanted to comment on the fact that not all sources necessarily have to depend on eventing so it might make sense to have a way to get the stats in from here.

**Comment by kthakar1990** [comment]
@evankanderson is this ready to merge?

File: source/source_stats_reporter.go [hunk 6]
```diff
@@ -98,6 +106,15 @@ func (r *reporter) ReportEventCount(args *ReportArgs, responseCode int) error {
 	return nil
 }
 
+func (r *reporter) ReportRetryEventCount(args *ReportArgs, responseCode int) error {
+	ctx, err := r.generateTag(args, responseCode)
+	if err != nil {
+		return err
+	}
+	metrics.Record(ctx, retryEventCountM.M(1))
```
Comment (n3wscott) [inline source/source_stats_reporter.go:114]: retry can happen more than once per attempt. I have not seen how this is intended to be used

Comment (kthakar1990) [reply source/source_stats_reporter.go:114]: you're right.. it happens more than once per attempt.. so the related PR: https://github.com/knative/eventing/pull/4861 is submitted in eventing repo.. that will call this reporter for every retry and that's why I am adding count here.

Comment (kthakar1990) [reply source/source_stats_reporter.go:114]: @n3wscott does that help?

Comment (vaikas) [reply source/source_stats_reporter.go:114]: Yeah, that makes sense to me. Otherwise you'd have to pass in something like an array of structs since if you collate the retry count with all the responseCodes that caused the failure, that seems wrong? Or, maybe I misunderstand what the question here was.

**Review by n3wscott — APPROVED** [review]
/lgtm
/approve
